### PR TITLE
Fix code gen for pattern switch with constant expression

### LINKED ISSUE #17090: Switch expression instrumentation not produced for constant expression in pattern switch
Code generation for a pattern switch does not evaluate the switch expression at runtime when it is a constant. Consequently there is no place in the generated code for the compiler to generate the instrumentation, and the compiler does not produce instrumentation for the switch expression in this case. We should evaluate what the proper treatment of this situation should be.

### LINKED ISSUE #19731: Bad code for pattern switch with constant expression
The following produces code that does nothing.

This is a regression in master versus the most recent update.

``` c#
    static void M1()
    {
        switch (1)
        {
            case 0 when true:
                ;
            case 1:
                Console.Write(1);
                break;
            case 2:
                ;
        }
    }
```

### LINKED ISSUE #19734: Incorrect code for generic pattern switch on a constant expression
The following produces incorrect code. When `T` is `long`, it prints nothing.

This is a regression in master versus the most recent update.

``` c#
    static void M1<T>()
    {
        switch (1)
        {
            case T t:
                Console.Write(1);
                break;
            case int i:
                Console.Write(2);
                break;
        }
    }
```

This is a new issue in 15.3, as the code was illegal previously.

### PR DESCRIPTION
**Pull request by gafter**
**Customer scenario**

Use new pattern-based switch constructs in which the switch expression is a constant.

**Bugs this fixes:**

* Fix code gen for pattern switch with constant expression
Fixes #19731

* Add instrumentation for pattern switch with constant expression
Fixes #17090

* Incorrect code for generic pattern switch on a constant expression
Fixes #19734

**Workarounds, if any**

Don't use a constant switch expression.

**Risk**

Low. The scenarios are fairly focused, and regression tests now exist for the problematic scenarios.

**Performance impact**

None expected. No significant change in the complexity of the relevant compiler code.

**Is this a regression from a previous update?**

Yes, #19731 is a regression from the last update.

**Root cause analysis:**

We are catching this now as we add coverage for the recently added support for switching on generic types.

**How was the bug found?**

Development testing.

### CODE REVIEW

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this bug fix for 15.3?

File: src/Compilers/CSharp/Test/Emit/PDB/PDBTests.cs [hunk 1]
```diff
@@ -2984,6 +2984,549 @@ class Student : Person { public double GPA; }
...
+  IL_0001:  ldc.i4.1
+  IL_0002:  stloc.0
+  IL_0003:  br.s       IL_0005
+  IL_0005:  ldc.i4.1
+  IL_0006:  call       ""void System.Console.Write(int)""
+  IL_000b:  nop
+  IL_000c:  br.s       IL_000e
+  IL_000e:  ret
+}");
+            c.VerifyPdb("Program.M1",
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/PDB/PDBTests.cs]: If your purpose is to verify sequence points, you could leverage VerifyIL to do it.
See options used in VerfiyIL in test `UsingStatement_EmbeddedWhile`.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 2]
```diff
@@ -248,61 +255,64 @@ private void LowerDecisionTree(BoundExpression expression, DecisionTree decision
...
+                                break;
+                            }
+                        case null:
+                            {
+                                // Unknown if the input is null. First test for null
+                                var notNullLabel = _factory.GenerateLabel("notNull");
+                                var inputExpression = byType.Expression;
+                                var objectType = _factory.SpecialType(SpecialType.System_Object);
+                                var nullValue = _factory.Null(objectType);
+                                BoundExpression notNull = byType.Type.IsNullableType() ? _localRewriter.RewriteNullableNullEquality(_factory.Syntax, BinaryOperatorKind.NullableNullNotEqual, byType.Expression, nullValue, _factory.SpecialType(SpecialType.System_Boolean))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: Consider breaking this line again (it's very long).

File: src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs [hunk 3]
```diff
@@ -308,8 +308,17 @@ private DecisionTree AddByValue(DecisionTree.ByType byType, BoundConstantPattern
             if (forType == null)
             {
                 var type = value.Value.Type;
-                var narrowedExpression = GetBoundPatternMatchingLocal(type);
-                forType = new DecisionTree.ByValue(narrowedExpression, type.TupleUnderlyingTypeOrSelf(), narrowedExpression.LocalSymbol);
+                if (byType.Type.Equals(type, TypeCompareKind.AllIgnoreOptions))
+                {
+                    // reuse the input expression when we can
+                    forType = new DecisionTree.ByValue(byType.Expression, type.TupleUnderlyingTypeOrSelf(), null);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs]: >forType = new DecisionTree.ByValue(byType.Expression, type.TupleUnderlyingTypeOrSelf(), null); [](start = 20, length = 94)

Is this change just an optimization, or is it needed for correctness? If the latter, please add a comment explaining why is this the case.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 4]
```diff
@@ -60,13 +60,20 @@ private BoundStatement MakeLoweredForm(BoundPatternSwitchStatement node)
...
+                    if (expression.ConstantValue == null)
                     {
-                        expression = _localRewriter._instrumenter.InstrumentSwitchStatementExpression(node, expression, _factory);
+                        expression = instrumentedExpression;
+                    }
+                    else
+                    {
+                        // If the expression is a constant, we leave it alone (the decision tree lowering code needs
+                        // to see that constant). But we add an additional leading statement with the instrumented expression.
+                        result.Add(_factory.ExpressionStatement(instrumentedExpression));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: >result.Add(_factory.ExpressionStatement(instrumentedExpression)); [](start = 24, length = 65)

It feels like we should add a test for instrumentation.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 5]
```diff
@@ -248,61 +255,64 @@ private void LowerDecisionTree(BoundExpression expression, DecisionTree decision
...
-                            LowerDecisionTree(byType.Expression, byType.Default);
-                        }
-                    }
-                    else
+                    inputIsNull = inputConstant.IsNull;
+                }
+
+                var defaultLabel = _factory.GenerateLabel("byTypeDefault");
+
+                if (byType.Type.CanContainNull())
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: >if (byType.Type.CanContainNull()) [](start = 16, length = 33)

Do we have test coverage for all branches in this ```if``` and the ```switch``` inside?

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs]: It is an optimization.

Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs]: Will make the comment more clear.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: Only the debug instrumentation is affected (it inserts a hidden sequence point). The other instrumentation forms ignore the switch expression. In other words, there is no instrumentation test that would observe a change in behavior.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: Yes.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: Yes.

File: src/Compilers/CSharp/Test/Emit/PDB/PDBTests.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Emit/PDB/PDBTests.cs]: Will do.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: In the case of a constant switch expression, no hidden sequence point is inserted, but instead an extra copy is inserted by the debug instrumentation code.

**Comment by gafter** [comment]
@jcouv @AlekseyTs The PR has been amended to address all comments. Do you have any further comments?

**Comment by gafter** [comment]
@jcouv Do you have any further comments?

**Comment by gafter** [comment]
retest perf_correctness_prtest please

(There was an I/O error during the build.)

**Comment by gafter** [comment]
retest perf_correctness_prtest please

**Comment by gafter** [comment]
@MeiChin-Tsai @jaredpar for ask-mode approval, please

**Comment by MeiChin-Tsai** [comment]
approved. thx.

**Comment by jcouv** [comment]
Looks good to me, but my level of familiarity in this area is limited. I pinged Andy to also take a look.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 6]
```diff
@@ -248,61 +261,71 @@ private void LowerDecisionTree(BoundExpression expression, DecisionTree decision
             private void LowerDecisionTree(DecisionTree.ByType byType)
             {
                 var inputConstant = byType.Expression.ConstantValue;
+
+                // three-valued: true if input known null, false if input known non-null, null if not known.
+                bool? inputIsNull = null;
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: Equivalent to `inputIsNull = inputConstant?.IsNull`?

File: src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs [hunk 7]
```diff
@@ -308,8 +308,17 @@ private DecisionTree AddByValue(DecisionTree.ByType byType, BoundConstantPattern
             if (forType == null)
             {
                 var type = value.Value.Type;
-                var narrowedExpression = GetBoundPatternMatchingLocal(type);
-                forType = new DecisionTree.ByValue(narrowedExpression, type.TupleUnderlyingTypeOrSelf(), narrowedExpression.LocalSymbol);
+                if (byType.Type.Equals(type, TypeCompareKind.AllIgnoreOptions))
+                {
+                    // reuse the input expression when we have an equivalent type to reduce the number of generated temps
+                    forType = new DecisionTree.ByValue(byType.Expression, type.TupleUnderlyingTypeOrSelf(), null);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs]: Yes, but that parameter is non-null only when it represents a (possibly) newly allocated temp (so we can keep track of which temps exist in the final decision tree). We are reusing the expression, so we are not allocating a new temp at this node.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs (hunk 6 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: Yes, that would have been simpler. I'll take a note to change it later.

**Comment by gafter** [comment]
@agocke Do you have any further comments on this?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs (hunk 6 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: Noted in https://github.com/dotnet/roslyn/issues/20029

**Pull request merged**